`.env` - Extra production protections & move injection from buildtime > runtime (#5162)
guardian/dotcom-rendering: dotcom-rendering/docs/architecture/028-dotenv.md

    // ...
];
```

## Security

`.env` variables are only available on the server, you should ensure that you **never** pass an environment variable into the clientside code
e.g

> ```jsx
> <Island>
> 	<!-- Bad :((( -->
> 	<MyComponent myProp={process.env.MY_VAR} />
> </Island>
> ```

Reviewer: Maybe we could give a good example here e.g. using the env var to pass a salted URL to the component?

What happens if we read a process.env in an `.importable`?

```suggestion
```jsx
return (
    <Island>
	    <!-- Bad :((( -->
	    <Image imageSalt={process.env.IMAGE_SALT} />
    </Island>
)

const url = getSaltedUrl(process.env.IMAGE_SALT)
return (
    <Island>
	    <!-- Good :))) -->
	    <Image url={url} />
    </Island>
)
```
```

Author: > What happens if we read a process.env in an .importable

On the client this will simply be undefined (and perhaps cause an error) and on the server is should work as expected

Reviewer: Worth adding the good example so people know how they might do it?

This has considerable security implementations as this variable would then be available to any consumer of the site.

> ## Limitations

Reviewer: Do we have any here?

Author: I'm not sure I understand the question!

Reviewer: Apologies, I didn't realise that this is a heading for the next points...

### Server Only

> DCR currently only supports `.env` variables being used on the _server_, but not the client.

Reviewer: nit: 
```suggestion
DCR currently only supports `.env` variables being used on the _server_, not the client.
```

> This limitation exists as the to maintain common & best practice within the organisation, we want to ensure that a re-deploy of the application would involve fetching a fresh environment variable. This means we cannot [inject environment variables through webpack](https://www.npmjs.com/package/dotenv-webpack) (which is done at build time) - and instead we load the dotenv directly when running the server bundle.

Reviewer: Does the fact we only make this available on the server ensure that we can make changes to env vars on redeploy? Or is it that we run this on start, not build?

Author: Both are necessary for this to be possible on redeploy. To make env variables available on the client you need to [inject them through webpack](https://github.com/mrsteele/dotenv-webpack#resulting-bundle), which happens at build time - moving to loading them with the regular `dotenv` package enabled us to use gen-env on start rather than build.

Reviewer: Ah, OK I see. I suppose I am getting confused as there are a few points:
* We don't want to expose these values to the client for security reasons (all are considered secrets)
* We want to have these values recyclable on redeploy

I wonder if these aren't limitations but what we expect, and should expect going forward?

    // ...
];
```

## Security

> `.env` variables are only available on the server, you should ensure that you **never** pass an environment variable into the clientside code

Reviewer 2: ```suggestion
`.env` variables are only available on the server, you should ensure that you **never** pass an environment variable into the client-side code.
```
